From cca91b7e7e480f406dfedc9081485b4598c98542 Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@t-sys.eu>
Date: Mon, 22 Mar 2021 00:51:13 +0100
Subject: [PATCH 11/11] ASoC: soc: codec: tas5825m: add left/right channel
 remapping controls

This commit adds controls for remapping left/right DAC inputs

Signed-off-by: Tobias Schramm <t.schramm@t-sys.eu>
---
 sound/soc/codecs/tas5825m.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/sound/soc/codecs/tas5825m.c b/sound/soc/codecs/tas5825m.c
index 314410f78042..9679fdcb71bf 100644
--- a/sound/soc/codecs/tas5825m.c
+++ b/sound/soc/codecs/tas5825m.c
@@ -552,7 +552,22 @@ static const struct regmap_config tas5825m_regmap_config = {
 	.volatile_reg = tas5825m_is_volatile_reg,
 };
 
-static const SNDRV_CTL_TLVD_DECLARE_DB_LINEAR(pga_gain_tlv, SNDRV_CTL_TLVD_DB_SCALE_MUTE, 2400);
+static const char *const left_dac_txt[] =
+	{ "Mute", "Left channel", "Right channel" };
+
+static const struct soc_enum left_dac_src =
+	SOC_ENUM_SINGLE(TAS5825M_SAP_CTRL3, 4, ARRAY_SIZE(left_dac_txt),
+			left_dac_txt);
+
+static const char *const right_dac_txt[] =
+	{ "Mute", "Right channel", "Left channel" };
+
+static const struct soc_enum right_dac_src =
+	SOC_ENUM_SINGLE(TAS5825M_SAP_CTRL3, 0, ARRAY_SIZE(right_dac_txt),
+			right_dac_txt);
+
+static const SNDRV_CTL_TLVD_DECLARE_DB_LINEAR(pga_gain_tlv,
+					SNDRV_CTL_TLVD_DB_SCALE_MUTE, 2400);
 
 static const SNDRV_CTL_TLVD_DECLARE_DB_LINEAR(again_tlv, -1550, 0);
 
@@ -561,6 +576,8 @@ static const struct snd_kcontrol_new tas5825m_snd_controls[] = {
 		       TAS5825M_DIG_VOL, 0, 0xff, 1, pga_gain_tlv),
 	SOC_SINGLE_TLV("Speaker Driver Analog Gain", TAS5825M_AGAIN,
 		       0, 0x1f, 1, again_tlv),
+	SOC_ENUM("Left DAC source", left_dac_src),
+	SOC_ENUM("Right DAC source", right_dac_src),
 };
 
 static const struct snd_soc_dapm_widget tas5825m_dapm_widgets[] = {
-- 
2.30.1

